{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "initial_id",
   "metadata": {
    "collapsed": true,
    "ExecuteTime": {
     "end_time": "2024-01-05T15:53:26.183942Z",
     "start_time": "2024-01-05T15:53:19.341210Z"
    }
   },
   "outputs": [],
   "source": [
    "import logging\n",
    "import operator\n",
    "from dataclasses import dataclass\n",
    "from enum import Enum\n",
    "from pathlib import Path\n",
    "\n",
    "import pandas as pd\n",
    "\n",
    "info_log = logging.getLogger(\"info\")\n",
    "\n",
    "\n",
    "def calculate_end_pos(variant_start: int, variant_ref: str) -> int:\n",
    "    \"\"\"Calculate the end position for a variant\n",
    "    Args:\n",
    "        variant_start (int): The start position of the variant\n",
    "        variant_ref (str): The reference allele of the variant\n",
    "\n",
    "    Returns:\n",
    "        int: The end position of the variant\n",
    "    \"\"\"\n",
    "    return variant_start + len(variant_ref) - 1\n",
    "\n",
    "\n",
    "@dataclass\n",
    "class PhEvalResult:\n",
    "    \"\"\"Base class for PhEval results.\"\"\"\n",
    "\n",
    "\n",
    "@dataclass\n",
    "class PhEvalGeneResult(PhEvalResult):\n",
    "    \"\"\"Minimal data required from tool-specific output for gene prioritisation result\n",
    "    Args:\n",
    "        gene_symbol (str): The gene symbol for the result entry\n",
    "        gene_identifier (str): The ENSEMBL gene identifier for the result entry\n",
    "        score (float): The score for the gene result entry\n",
    "    Notes:\n",
    "        While we recommend providing the gene identifier in the ENSEMBL namespace,\n",
    "        any matching format used in Phenopacket interpretations is acceptable for result matching purposes\n",
    "        in the analysis.\n",
    "    \"\"\"\n",
    "\n",
    "    gene_symbol: str\n",
    "    gene_identifier: str\n",
    "    score: float\n",
    "\n",
    "\n",
    "@dataclass\n",
    "class RankedPhEvalGeneResult(PhEvalGeneResult):\n",
    "    \"\"\"PhEval gene result with corresponding rank\n",
    "    Args:\n",
    "        rank (int): The rank for the result entry\n",
    "    \"\"\"\n",
    "\n",
    "    rank: int\n",
    "\n",
    "    @staticmethod\n",
    "    def from_gene_result(pheval_gene_result: PhEvalGeneResult, rank: int):\n",
    "        \"\"\"Return RankedPhEvalGeneResult from a PhEvalGeneResult and rank\n",
    "        Args:\n",
    "            pheval_gene_result (PhEvalGeneResult): The gene result entry\n",
    "            rank (int): The corresponding rank for the result entry\n",
    "\n",
    "        Returns:\n",
    "            RankedPhEvalGeneResult: The result as a RankedPhEvalGeneResult\n",
    "        \"\"\"\n",
    "        return RankedPhEvalGeneResult(\n",
    "            gene_symbol=pheval_gene_result.gene_symbol,\n",
    "            gene_identifier=pheval_gene_result.gene_identifier,\n",
    "            score=pheval_gene_result.score,\n",
    "            rank=rank,\n",
    "        )\n",
    "\n",
    "\n",
    "@dataclass\n",
    "class PhEvalVariantResult(PhEvalResult):\n",
    "    \"\"\"Minimal data required from tool-specific output for variant prioritisation\n",
    "    Args:\n",
    "        chromosome (str): The chromosome position of the variant recommended to be provided in the following format.\n",
    "        This includes numerical designations from 1 to 22 representing autosomal chromosomes,\n",
    "        as well as the sex chromosomes X and Y, and the mitochondrial chromosome MT.\n",
    "        start (int): The start position of the variant\n",
    "        end (int): The end position of the variant\n",
    "        ref (str): The reference allele of the variant\n",
    "        alt (str): The alternate allele of the variant\n",
    "        score (float): The score for the variant result entry\n",
    "    Notes:\n",
    "        While we recommend providing the variant's chromosome in the specified format,\n",
    "        any matching format used in Phenopacket interpretations is acceptable for result matching purposes\n",
    "        in the analysis.\n",
    "    \"\"\"\n",
    "\n",
    "    chromosome: str\n",
    "    start: int\n",
    "    end: int\n",
    "    ref: str\n",
    "    alt: str\n",
    "    score: float\n",
    "\n",
    "\n",
    "@dataclass\n",
    "class RankedPhEvalVariantResult(PhEvalVariantResult):\n",
    "    \"\"\"PhEval variant result with corresponding rank\n",
    "    Args:\n",
    "        rank (int): The rank for the result entry\n",
    "    \"\"\"\n",
    "\n",
    "    rank: int\n",
    "\n",
    "    @staticmethod\n",
    "    def from_variant_result(pheval_variant_result: PhEvalVariantResult, rank: int):\n",
    "        \"\"\"Return RankedPhEvalVariantResult from a PhEvalVariantResult and rank\n",
    "        Args:\n",
    "            pheval_variant_result (PhEvalVariantResult): The variant result entry\n",
    "            rank (int): The corresponding rank for the result entry\n",
    "\n",
    "        Returns:\n",
    "            RankedPhEvalVariantResult: The result as a RankedPhEvalVariantResult\n",
    "        \"\"\"\n",
    "        return RankedPhEvalVariantResult(\n",
    "            chromosome=pheval_variant_result.chromosome,\n",
    "            start=pheval_variant_result.start,\n",
    "            end=pheval_variant_result.end,\n",
    "            ref=pheval_variant_result.ref,\n",
    "            alt=pheval_variant_result.alt,\n",
    "            score=pheval_variant_result.score,\n",
    "            rank=rank,\n",
    "        )\n",
    "\n",
    "\n",
    "@dataclass\n",
    "class PhEvalDiseaseResult(PhEvalResult):\n",
    "    \"\"\"Minimal data required from tool-specific output for disease prioritisation\n",
    "    Args:\n",
    "        disease_name (str): Disease name for the result entry\n",
    "        disease_identifier (str): Identifier for the disease result entry in the OMIM namespace\n",
    "        score (str): Score for the disease result entry\n",
    "    Notes:\n",
    "        While we recommend providing the disease identifier in the OMIM namespace,\n",
    "        any matching format used in Phenopacket interpretations is acceptable for result matching purposes\n",
    "        in the analysis.\n",
    "    \"\"\"\n",
    "\n",
    "    disease_name: str\n",
    "    disease_identifier: str\n",
    "    score: float\n",
    "\n",
    "\n",
    "@dataclass\n",
    "class RankedPhEvalDiseaseResult(PhEvalDiseaseResult):\n",
    "    \"\"\"PhEval disease result with corresponding rank\n",
    "    Args:\n",
    "        rank (int): The rank for the result entry\n",
    "    \"\"\"\n",
    "\n",
    "    rank: int\n",
    "\n",
    "    @staticmethod\n",
    "    def from_disease_result(pheval_disease_result: PhEvalDiseaseResult, rank: int):\n",
    "        \"\"\"Return RankedPhEvalDiseaseResult from a PhEvalDiseaseResult and rank\n",
    "        Args:\n",
    "            pheval_disease_result (PhEvalDiseaseResult): The disease result entry\n",
    "            rank (int): The corresponding rank for the result entry\n",
    "\n",
    "        Returns:\n",
    "            RankedPhEvalDiseaseResult: The result as a RankedPhEvalDiseaseResult\n",
    "        \"\"\"\n",
    "        return RankedPhEvalDiseaseResult(\n",
    "            disease_name=pheval_disease_result.disease_name,\n",
    "            disease_identifier=pheval_disease_result.disease_identifier,\n",
    "            score=pheval_disease_result.score,\n",
    "            rank=rank,\n",
    "        )\n",
    "\n",
    "\n",
    "class SortOrder(Enum):\n",
    "    \"\"\"Enumeration representing sorting orders.\"\"\"\n",
    "\n",
    "    ASCENDING = 1\n",
    "    \"\"\"Ascending sort order.\"\"\"\n",
    "    DESCENDING = 2\n",
    "    \"\"\"Descending sort order.\"\"\"\n",
    "\n",
    "\n",
    "class ResultSorter:\n",
    "    \"\"\"Class for sorting PhEvalResult instances based on a given sort order.\"\"\"\n",
    "\n",
    "    def __init__(self, pheval_results: [PhEvalResult], sort_order: SortOrder):\n",
    "        \"\"\"\n",
    "        Initialise ResultSorter\n",
    "\n",
    "        Args:\n",
    "            pheval_results ([PhEvalResult]): List of PhEvalResult instances to be sorted\n",
    "            sort_order (SortOrder): Sorting order to be applied\n",
    "        \"\"\"\n",
    "        self.pheval_results = pheval_results\n",
    "        self.sort_order = sort_order\n",
    "\n",
    "    def _sort_by_decreasing_score(self) -> [PhEvalResult]:\n",
    "        \"\"\"\n",
    "        Sort results in descending order based on the score\n",
    "\n",
    "        Returns:\n",
    "            [PhEvalResult]: Sorted list of PhEvalResult instances.\n",
    "        \"\"\"\n",
    "        return sorted(self.pheval_results, key=operator.attrgetter(\"score\"), reverse=True)\n",
    "\n",
    "    def _sort_by_increasing_score(self) -> [PhEvalResult]:\n",
    "        \"\"\"\n",
    "        Sort results in ascending order based on the score\n",
    "\n",
    "        Returns:\n",
    "            [PhEvalResult]: Sorted list of PhEvalResult instances.\n",
    "        \"\"\"\n",
    "        return sorted(self.pheval_results, key=operator.attrgetter(\"score\"), reverse=False)\n",
    "\n",
    "    def sort_pheval_results(self) -> [PhEvalResult]:\n",
    "        \"\"\"\n",
    "        Sort results based on the specified sort order.\n",
    "\n",
    "        Returns:\n",
    "            [PhEvalResult]: Sorted list of PhEvalResult instances.\n",
    "        \"\"\"\n",
    "        return (\n",
    "            self._sort_by_increasing_score()\n",
    "            if self.sort_order == SortOrder.ASCENDING\n",
    "            else self._sort_by_decreasing_score()\n",
    "        )\n",
    "\n",
    "\n",
    "class ScoreRanker:\n",
    "    \"\"\"\n",
    "    Class for ranking scores based on a given sort order\n",
    "\n",
    "    Attributes:\n",
    "       rank (int): Represents the current rank, initialised with 0\n",
    "       current_score (float): Represents the current score, initialised with positive infinity (float(\"inf\"))\n",
    "       count (int): Used for counting, initialised with 0\n",
    "    \"\"\"\n",
    "\n",
    "    rank: int = 0\n",
    "    current_score: float = float(\"inf\")\n",
    "    count: int = 0\n",
    "\n",
    "    def __init__(self, sort_order: SortOrder):\n",
    "        \"\"\"\n",
    "        Initialise ScoreRanker\n",
    "\n",
    "        Args:\n",
    "            sort_order (SortOrder): Sorting order to be applied\n",
    "        \"\"\"\n",
    "        self.sort_order = sort_order\n",
    "\n",
    "    def _check_rank_order(self, round_score: float) -> None:\n",
    "        \"\"\"\n",
    "        Check if the results are correctly ordered\n",
    "\n",
    "        Args:\n",
    "            round_score (float): Score to be checked against the current score\n",
    "\n",
    "        Raises:\n",
    "            ValueError: If results are not correctly sorted.\n",
    "        \"\"\"\n",
    "        if self.sort_order == SortOrder.ASCENDING and round_score < self.current_score != float(\n",
    "            \"inf\"\n",
    "        ):\n",
    "            raise ValueError(\"Results are not correctly sorted!\")\n",
    "        elif self.sort_order == SortOrder.DESCENDING and round_score > self.current_score != float(\n",
    "            \"inf\"\n",
    "        ):\n",
    "            raise ValueError(\"Results are not correctly sorted!\")\n",
    "\n",
    "    def rank_scores(self, round_score: float) -> int:\n",
    "        \"\"\"\n",
    "        Add ranks to a result; equal scores are given the same rank, e.g., 1, 1, 3\n",
    "\n",
    "        Args:\n",
    "            round_score (float): Score to be ranked\n",
    "\n",
    "        Returns:\n",
    "            int: Rank assigned to the score\n",
    "        \"\"\"\n",
    "        self._check_rank_order(round_score)\n",
    "        self.count += 1\n",
    "        if self.current_score == round_score:\n",
    "            return self.rank\n",
    "        self.current_score = round_score\n",
    "        self.rank = self.count\n",
    "        return self.rank\n",
    "\n",
    "\n",
    "def _rank_pheval_result(pheval_result: [PhEvalResult], sort_order: SortOrder) -> [PhEvalResult]:\n",
    "    \"\"\"\n",
    "    Rank PhEval results post-processed from tool-specific output, managing tied scores (ex aequo)\n",
    "\n",
    "    Args:\n",
    "        pheval_result ([PhEvalResult]): PhEval results obtained from tool-specific output\n",
    "        sort_order (SortOrder): Sorting order based on which ranking is performed\n",
    "\n",
    "    Returns:\n",
    "        List[PhEvalResult]: Ranked PhEval results with tied scores managed\n",
    "\n",
    "    Raises:\n",
    "        ValueError: If an incompatible PhEval result type is encountered\n",
    "    \"\"\"\n",
    "    score_ranker = ScoreRanker(sort_order)\n",
    "    ranked_result = []\n",
    "    for result in pheval_result:\n",
    "        if type(result) == PhEvalGeneResult:\n",
    "            ranked_result.append(\n",
    "                RankedPhEvalGeneResult.from_gene_result(\n",
    "                    result, score_ranker.rank_scores(result.score)\n",
    "                )\n",
    "            )\n",
    "        elif type(result) == PhEvalVariantResult:\n",
    "            ranked_result.append(\n",
    "                RankedPhEvalVariantResult.from_variant_result(\n",
    "                    result, score_ranker.rank_scores(result.score)\n",
    "                )\n",
    "            )\n",
    "        elif type(result) == PhEvalDiseaseResult:\n",
    "            ranked_result.append(\n",
    "                RankedPhEvalDiseaseResult.from_disease_result(\n",
    "                    result, score_ranker.rank_scores(result.score)\n",
    "                )\n",
    "            )\n",
    "        else:\n",
    "            raise ValueError(\"Incompatible PhEval result type.\")\n",
    "    return ranked_result\n",
    "\n",
    "\n",
    "def _return_sort_order(sort_order_str: str) -> SortOrder:\n",
    "    \"\"\"\n",
    "    Convert a string derived from the config file into SortOrder Enum\n",
    "\n",
    "    Args:\n",
    "        sort_order_str (str): String representation of the sorting order\n",
    "\n",
    "    Returns:\n",
    "        SortOrder: Enum representing the specified sorting order\n",
    "\n",
    "    Raises:\n",
    "        ValueError: If an incompatible or unknown sorting method is provided\n",
    "    \"\"\"\n",
    "    try:\n",
    "        return SortOrder[sort_order_str.upper()]\n",
    "    except KeyError:\n",
    "        raise ValueError(\"Incompatible ordering method specified.\")\n",
    "\n",
    "\n",
    "def _create_pheval_result(pheval_result: [PhEvalResult], sort_order_str: str) -> [PhEvalResult]:\n",
    "    \"\"\"\n",
    "    Create PhEval results with corresponding ranks based on the specified sorting order.\n",
    "\n",
    "    Args:\n",
    "        pheval_result ([PhEvalResult]): List of PhEvalResult instances to be processed.\n",
    "        sort_order_str (str): String representation of the desired sorting order.\n",
    "\n",
    "    Returns:\n",
    "        List[PhEvalResult]: PhEval results with ranks assigned.\n",
    "    \"\"\"\n",
    "    sort_order = _return_sort_order(sort_order_str)\n",
    "    sorted_pheval_result = ResultSorter(pheval_result, sort_order).sort_pheval_results()\n",
    "    return _rank_pheval_result(sorted_pheval_result, sort_order)\n",
    "\n",
    "\n",
    "def _write_pheval_gene_result(\n",
    "    ranked_pheval_result: [PhEvalResult], output_dir: Path, tool_result_path: Path\n",
    ") -> None:\n",
    "    \"\"\"\n",
    "    Write ranked PhEval gene results to a TSV file\n",
    "\n",
    "    Args:\n",
    "        ranked_pheval_result ([PhEvalResult]): List of ranked PhEval gene results\n",
    "        output_dir (Path): Path to the output directory\n",
    "        tool_result_path (Path): Path to the tool-specific result file\n",
    "    \"\"\"\n",
    "    ranked_result = pd.DataFrame([data.__dict__ for data in ranked_pheval_result])\n",
    "    pheval_gene_output = ranked_result.loc[:, [\"rank\", \"score\", \"gene_symbol\", \"gene_identifier\"]]\n",
    "    pheval_gene_output.to_csv(\n",
    "        output_dir.joinpath(\n",
    "            \"pheval_gene_results/\" + tool_result_path.stem + \"-pheval_gene_result.tsv\"\n",
    "        ),\n",
    "        sep=\"\\t\",\n",
    "        index=False,\n",
    "    )\n",
    "\n",
    "\n",
    "def _write_pheval_variant_result(\n",
    "    ranked_pheval_result: [PhEvalResult], output_dir: Path, tool_result_path: Path\n",
    ") -> None:\n",
    "    \"\"\"\n",
    "    Write ranked PhEval variant results to a TSV file\n",
    "\n",
    "    Args:\n",
    "        ranked_pheval_result ([PhEvalResult]): List of ranked PhEval gene results\n",
    "        output_dir (Path): Path to the output directory\n",
    "        tool_result_path (Path): Path to the tool-specific result file\n",
    "    \"\"\"\n",
    "    ranked_result = pd.DataFrame([data.__dict__ for data in ranked_pheval_result])\n",
    "    pheval_variant_output = ranked_result.loc[\n",
    "        :, [\"rank\", \"score\", \"chromosome\", \"start\", \"end\", \"ref\", \"alt\"]\n",
    "    ]\n",
    "    pheval_variant_output.to_csv(\n",
    "        output_dir.joinpath(\n",
    "            \"pheval_variant_results/\" + tool_result_path.stem + \"-pheval_variant_result.tsv\"\n",
    "        ),\n",
    "        sep=\"\\t\",\n",
    "        index=False,\n",
    "    )\n",
    "\n",
    "\n",
    "def _write_pheval_disease_result(\n",
    "    ranked_pheval_result: [RankedPhEvalDiseaseResult], output_dir: Path, tool_result_path: Path\n",
    ") -> None:\n",
    "    \"\"\"\n",
    "    Write ranked PhEval disease results to a TSV file\n",
    "\n",
    "    Args:\n",
    "        ranked_pheval_result ([PhEvalResult]): List of ranked PhEval gene results\n",
    "        output_dir (Path): Path to the output directory\n",
    "        tool_result_path (Path): Path to the tool-specific result file\n",
    "    \"\"\"\n",
    "    ranked_result = pd.DataFrame([data.__dict__ for data in ranked_pheval_result])\n",
    "    pheval_disease_output = ranked_result.loc[\n",
    "        :, [\"rank\", \"score\", \"disease_name\", \"disease_identifier\"]\n",
    "    ]\n",
    "    pheval_disease_output.to_csv(\n",
    "        output_dir.joinpath(\n",
    "            \"pheval_disease_results/\" + tool_result_path.stem + \"-pheval_disease_result.tsv\"\n",
    "        ),\n",
    "        sep=\"\\t\",\n",
    "        index=False,\n",
    "    )\n",
    "\n",
    "\n",
    "def generate_pheval_result(\n",
    "    pheval_result: [PhEvalResult],\n",
    "    sort_order_str: str,\n",
    "    output_dir: Path,\n",
    "    tool_result_path: Path,\n",
    ") -> None:\n",
    "    \"\"\"\n",
    "    Generate PhEval variant, gene or disease TSV result based on input results.\n",
    "\n",
    "    Args:\n",
    "        pheval_result ([PhEvalResult]): List of PhEvalResult instances to be processed.\n",
    "        sort_order_str (str): String representation of the desired sorting order.\n",
    "        output_dir (Path): Path to the output directory.\n",
    "        tool_result_path (Path): Path to the tool-specific result file.\n",
    "\n",
    "    Raises:\n",
    "        ValueError: If the results are not all the same type or an error occurs during file writing.\n",
    "    \"\"\"\n",
    "    if not pheval_result:\n",
    "        info_log.warning(f\"No results found for {tool_result_path.name}\")\n",
    "        pass\n",
    "    ranked_pheval_result = _create_pheval_result(pheval_result, sort_order_str)\n",
    "    if all(isinstance(result, RankedPhEvalGeneResult) for result in ranked_pheval_result):\n",
    "        _write_pheval_gene_result(ranked_pheval_result, output_dir, tool_result_path)\n",
    "    elif all(isinstance(result, RankedPhEvalVariantResult) for result in ranked_pheval_result):\n",
    "        _write_pheval_variant_result(ranked_pheval_result, output_dir, tool_result_path)\n",
    "    elif all(isinstance(result, RankedPhEvalDiseaseResult) for result in ranked_pheval_result):\n",
    "        _write_pheval_disease_result(ranked_pheval_result, output_dir, tool_result_path)\n",
    "    else:\n",
    "        raise ValueError(\"Results are not all of the same type.\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/Users/carlo/PycharmProjects/RAGexomiser/notebooks\r\n"
     ]
    }
   ],
   "source": [
    "!pwd"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2024-01-05T15:53:34.930775Z",
     "start_time": "2024-01-05T15:53:34.805880Z"
    }
   },
   "id": "e55d8697955744e9"
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "outputs": [],
   "source": [],
   "metadata": {
    "collapsed": false
   },
   "id": "f038f1147d5b2e06"
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
